Mathlib/Tactic/FProp/RefinedDiscrTree.lean

```diff
@@ -0,0 +1,844 @@
+`∀ x, p[x]` and `∃ x, p[x]` get the same size. -/
+partial def DTExpr.size : DTExpr → Nat
+| .const _ args
+| .fvar _ args
+| .bvar _ args => args.foldl (init := 1) (fun total a => total + a.size)
+| .lam b => b.size
+| .forall d b => 1 + d.size + b.size
+| _ => 1
+
+
+-- ## Encoding an Expr
+
+/-- This state is used to turn the indexing by `MVarId` and `FVarId` in `DTExpr` into
+indexing by `Nat` in `Key`. -/
+private structure Flatten.State where
+  stars : Array MVarId := #[]
+
+private def getStar (mvarId? : Option MVarId) : StateM Flatten.State Nat :=
+  modifyGet fun s =>
+    match mvarId? with
+    | some mvarId => match s.stars.findIdx? (· == mvarId) with
+      | some idx => (idx, s)
+      | none => (s.stars.size, { s with stars := s.stars.push mvarId })
+    | none => (s.stars.size, { s with stars := s.stars.push ⟨.anonymous⟩ })
+
+private partial def DTExpr.flattenAux (todo : Array Key) : DTExpr → StateM Flatten.State (Array Key)
+  | .star i => return todo.push (.star (← getStar i))
+  | .opaque => return todo.push .opaque
+  | .const n as => as.foldlM flattenAux (todo.push (.const n as.size))
+  | .fvar  f as => as.foldlM flattenAux (todo.push (.fvar f as.size))
+  | .bvar  i as => as.foldlM flattenAux (todo.push (.bvar i as.size))
+  | .lit l => return todo.push (.lit l)
+  | .sort  => return todo.push .sort
+  | .lam b => flattenAux (todo.push .lam) b
+  | .«forall» d b => do flattenAux (← flattenAux (todo.push .forall) d) b
+  | .proj n i e as => do as.foldlM flattenAux (← flattenAux (todo.push (.proj n i as.size)) e)
+
+/-- Given a `DTExpr`, return the linearized encoding in terms of `Key`,
+which is used for `RefinedDiscrTree` indexing. -/
+def DTExpr.flatten (e : DTExpr) (initCapacity := 16) : Array Key :=
+  (DTExpr.flattenAux (.mkEmpty initCapacity) e).run' {}
+
+
+
+/-- Return true if `a` should be ignored in the `RefinedDiscrTree`. -/
+private def ignoreArg (a : Expr) (i : Nat) (infos : Array ParamInfo) : MetaM Bool := do
+  if h : i < infos.size then
+    let info := infos.get ⟨i, h⟩
+    if info.isInstImplicit then
+      return true
+    else if info.isImplicit || info.isStrictImplicit then
+      return !(← isType a)
+    else
+      isProof a
+  else
+    isProof a
+
+/-- Return true if `e` is one of the following
+- A nat literal (numeral)
+- `Nat.zero`
+- `Nat.succ x` where `isNumeral x`
+- `OfNat.ofNat _ x _` where `isNumeral x` -/
+private partial def isNumeral (e : Expr) : Bool :=
+  if e.isNatLit then true
+  else
+    let f := e.getAppFn
+    if !f.isConst then false
+    else
+      let fName := f.constName!
+      if fName == ``Nat.succ && e.getAppNumArgs == 1 then isNumeral e.appArg!
+      else if fName == ``OfNat.ofNat && e.getAppNumArgs == 3 then isNumeral (e.getArg! 1)
+      else if fName == ``Nat.zero && e.getAppNumArgs == 0 then true
+      else false
+
+/-- Return `some n` if `e` is definitionally equal to the natural number `n`. -/
+private partial def toNatLit? (e : Expr) : Option Literal :=
```

Comment:
I'm a bit surprised that the logic is repeated with `isNumeral`. Is it really necessary that these are separate?

Reply:
If forgot to mention that `RefinedDiscrTree` and `StateList` are not really part of this PR. There is [std4#394](https://github.com/leanprover/std4/pulls?q=is%3Apr+is%3Aopen+tree) for it.